WcaOnRails/app/models/anonymize_person.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+class AnonymizePerson
+  include ActiveModel::Model
+
+  attr_writer :current_step
+  attr_reader :person_wca_id, :new_wca_id, :person, :account
+
+  def person_wca_id=(wca_id)
+    @person_wca_id = wca_id
+    @person = Person.find_by_wca_id(person_wca_id)
+    @account = User.find_by_wca_id(person_wca_id)
+  end
+
+  validates :person_wca_id, presence: true
+
+  def current_step
+    @current_step || steps.first
+  end
+
+  def steps
+    %w[enter_wca_id request_data_removal]
+  end
+
+  def next_step
```

Comment:
Awesome! Good to know. Is there a reason this is not done for do_anonymize_person? ie. It is not done for `do_merge` in `merge_people.rb`

Reply:
From briefly looking at the code, it seems like `do_merge` in that file doesn't actually mutate the object in the stricter sense. It modifies the database contents backing the Ruby models, but by the time you are calling that method on an object, said object doesn't immediately change.

```ruby
foo_person = Person.find(wca_id: 1234ABCD01)
puts foo_person.country # prints "Foonited States of Barmerica"

Person.where(wca_id: foo_person.wca_id).update(countryId: not_usa)
puts foo_person.country # **still prints "Foonited States of Barmerica"** because the Ruby object was retrieved at a time when country_id was not yet changed.

foo_person_again = Person.find(wca_id: 1234ABCD01)
puts foo_person_again.country # prints "Not the USA"

puts foo_person.countryId == foo_person_again.countryId # false, see above.
```

So as you can see, _technically_ `foo_person` has never been modified from Ruby's point of view, thus no exclamation mark

Reply:
On top of my overly verbose explanation, note that the exclamation mark is just a human convention and in no way enforced by Ruby. So it could really also depend on who wrote the code or who provided the Code Review.